### PR TITLE
Fix verification for successful staging

### PR DESCRIPTION
**Pull request by IvanBorislavovDimitrov**
Description:
Check for existing uploaded and valid packages and use them when the content of an application hasn't been changed. This way the app won't be restaged during the start which causes "app has not finished staging" exception.

Issue: LMCROSSITXSADEPLOY-2040

### CODE REVIEW

File: multiapps-controller-process/src/test/java/org/cloudfoundry/multiapps/controller/process/util/CloudPackagesGetterTest.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        assertEquals("500 Internal Server Error", exception.getMessage());
+    }
+
+    @Test
+    void getLatestUnusedPackageWithOneValidPackageNoDroplet() {
+        Mockito.when(client.getCurrentDropletForApplication(APPLICATION_GUID))
+               .thenThrow(getNotFoundCloudOperationException());
+        Mockito.when(client.getPackagesForApplication(APPLICATION_GUID))
+               .thenReturn(Collections.singletonList(createCloudPackage(PACKAGE_GUID, Status.PROCESSING_UPLOAD,
+                                                                        new Date(System.currentTimeMillis()))));
```
Comment (radoslav-d) [inline multiapps-controller-process/src/test/java/org/cloudfoundry/multiapps/controller/process/util/CloudPackagesGetterTest.java:62]: maybe `new Date()` is enough

File: multiapps-controller-process/src/test/java/org/cloudfoundry/multiapps/controller/process/util/CloudPackagesGetterTest.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+               .thenThrow(getNotFoundCloudOperationException());
+        Optional<CloudPackage> latestUnusedPackage = cloudPackagesGetter.getLatestUnusedPackage(client, APPLICATION_GUID);
+        assertFalse(latestUnusedPackage.isPresent());
+    }
+
+    @Test
+    void getLatestUnusedPackageWhenCurrentPackageIsTheSameAsNewestPackage() {
+        Mockito.when(client.getCurrentDropletForApplication(APPLICATION_GUID))
+               .thenReturn(createDropletInfo(DROPLET_GUID, PACKAGE_GUID));
+        CloudPackage cloudPackage = createCloudPackage(PACKAGE_GUID, Status.READY, new Date(System.currentTimeMillis()));
```
Comment (radoslav-d) [inline multiapps-controller-process/src/test/java/org/cloudfoundry/multiapps/controller/process/util/CloudPackagesGetterTest.java:83]: same `new Date()`

File: pom.xml [hunk 3]
```diff
@@ -52,7 +52,7 @@
         <netty-codec-http.version>4.1.49.Final</netty-codec-http.version>
         <liquibase.version>3.4.1</liquibase.version>
         <liquibase-slf4j.version>1.2.1</liquibase-slf4j.version>
-        <cloudfoundry.client.version>1.26.0</cloudfoundry.client.version>
+        <cloudfoundry.client.version>1.27.0-SNAPSHOT</cloudfoundry.client.version>
```
Comment (theghost5800) [inline pom.xml:55]: Do not forget to adopt release version

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/UploadAppStep.java [hunk 4]
```diff
@@ -43,52 +48,65 @@
...
+    private StepPhase useLatestPackage(ProcessContext context, CloudPackage latestUnusedPackage) {
+        getStepLogger().debug(Messages.THE_NEWEST_PACKAGE_WILL_BE_USED_0, SecureSerialization.toJson(latestUnusedPackage));
+        context.setVariable(Variables.UPLOAD_TOKEN, ImmutableUploadToken.builder()
+                                                                        .packageGuid(latestUnusedPackage.getGuid())
+                                                                        .build());
+        return StepPhase.POLL;
+    }
+
+    private boolean isCloudPackageInValidState(CloudPackage cloudPackage) {
+        return cloudPackage.getStatus() != Status.EXPIRED && cloudPackage.getStatus() != Status.FAILED;
```
[the remark this answers is not in the document]
Comment (IvanBorislavovDimitrov) [reply multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/UploadAppStep.java]: As we discussed only COPYING and PROCESSING_UPLOAD are safe enough.

File: multiapps-controller-process/src/test/java/org/cloudfoundry/multiapps/controller/process/steps/UploadAppStepTest.java [hunk 5]
```diff
@@ -187,72 +109,130 @@ public void prepareContext() {
...
+        void testWithAvailableFailedLatestPackageAndNonChangedApplicationContent() throws Exception {
+            prepareClients(null, null, CURRENT_MODULE_DIGEST);
+            mockCloudPackagesGetter(createCloudPackage(Status.FAILED));
+            step.execute(execution);
+            assertEquals(UPLOAD_TOKEN, context.getVariable(Variables.UPLOAD_TOKEN));
+            assertEquals(StepPhase.POLL.toString(), getExecutionStatus());
+        }
+
+        @Test
+        void testWithAvailableExpiredCloudPackageAndChangedContent() throws Exception {
```
[the remark this answers is not in the document]
Comment (IvanBorislavovDimitrov) [reply multiapps-controller-process/src/test/java/org/cloudfoundry/multiapps/controller/process/steps/UploadAppStepTest.java]: Done

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/UploadAppStep.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (IvanBorislavovDimitrov) [reply multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/UploadAppStep.java]: Until the application binary is not fully uploaded the current status is AWAITING_UPLOAD so, it is safe to poll PROCESSING_UPLOAD and COPYING statuses.

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/util/ApplicationStager.java [hunk 6]
```diff
@@ -36,8 +36,8 @@ public StagingState getStagingState() {
         UUID buildGuid = context.getVariable(Variables.BUILD_GUID);
         if (buildGuid == null) {
             return ImmutableStagingState.builder()
-                                        .state(PackageState.STAGED)
-                                        .build();
+                    .state(PackageState.STAGED)
```
Comment (theghost5800) [inline multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/util/ApplicationStager.java:39]: It looks like is not applied right formatter

Comment (IvanBorislavovDimitrov) [reply multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/util/ApplicationStager.java]: Fixed

**Pull request merged**